text/048-packaging-coq-native.md

```diff
@@ -0,0 +1,178 @@
+
+# Detailed design
+
+The proposal is three-fold:
+
+* **item 1.** Update the `coq` packaging in [ocaml/opam-repository](https://github.com/ocaml/opam-repository) so that Coq's standard library is compiled with `-native-compiler no` by default, unless an additional meta-package `coq-native` has been installed by the user.
+* **item 2.** Extend the `./configure -native-compiler` options:
+    * option `./configure -native-compiler yes` now impacts the default value of option `coqc -native-compiler` (in order to precompile both stdlib and third-party libraries with `-native-compiler yes`);
+	* option `-native-compiler ondemand` (which becomes the default when compiling coq manually) preserves the previous default behavior (modulo the stdlib that is not precompiled anymore).
+* **item 3.** Optionally: to enhance `native_compute` support with old versions of Coq (the releases of Coq before 8.13 where **item 2** is implemented), update the `opam` packaging of the considered libraries.
+
+
+## Regarding item 1:
+
+* Add an opam metapackage `coq-native.1`:
+
+```
+opam-version: "2.0"
+maintainer: "Erik Martin-Dorel"
+authors: "Coq"
+homepage: "https://coq.inria.fr/"
+bug-reports: "https://github.com/coq/coq/issues"
+conflicts: [
+  "coq" {< "8.5"}
+]
+synopsis: "Package flag enabling coq's native-compiler flag"
+description: """
+This package acts as a package flag for the ambient switch, taken into
+account by coq (and possibly any coq library) to enable native_compute
+at configure time, triggering the installation of .coq-native/* files
+for the coq standard library and subsequent coq libraries.
+
+This implements item 1 of CEP #48 <https://github.com/coq/ceps/pull/48>.
+
+Remarks:
+
+1. you might face with issues installing this package flag under macOS,
+   see <https://github.com/coq/coq/issues/11178>.
+2. this package is not intended to be used as a dependency of other
+   packages (notably as installing or uninstalling this package may
+   trigger a rebuild of all coq packages in the ambient switch).
+3. the option set by this package will be automatically propagated to
+   coqc for coq >= 8.13 (but the packaging of coq libraries for
+   earlier versions of coq may need an update: for details, see item 3
+   of CEP #48 <https://github.com/coq/ceps/pull/48>).
+"""
+```
+
+* Add the following lines in the opam packages of `coq`:
+
+```diff
+diff --git a/packages/coq/coq.8.11.2/opam b/packages/coq/coq.8.11.2/opam
+index fdb19e9da7..3d3dee7570 100644
+--- a/packages/coq/coq.8.11.2/opam
++++ b/packages/coq/coq.8.11.2/opam
+@@ -23,6 +23,9 @@ depends: [
+   "num"
+   "conf-findutils" {build}
+ ]
++depopts: [
++  "coq-native"
++]
+ build: [
+   [
+     "./configure"
+@@ -33,7 +36,9 @@ build: [
+     "-libdir" "%{lib}%/coq"
+     "-datadir" "%{share}%/coq"
+     "-coqide" "no"
+-    "-native-compiler" {os = "macos"} "no" {os = "macos"}
++    "-native-compiler" "yes" {coq-native:installed} "no" {!coq-native:installed}
```

Comment:
```suggestion
+    "-native-compiler" {coq-native:installed} "yes" {coq-native:installed}
```
so that native_compute remains available via ondemand when coq-native is not installed?

Reply:
Thanks @proux01 for all your suggestions, I will take them into account tonight.

> so that native_compute remains available via ondemand when coq-native is not installed?

This is the main question raised by your review.

Actually I had thought about this idea as well, but preferred to keep the CEP as is, given:
* the Coq call that had approved the version with `no coq-native package` → `configure -native-compiler no`,
* some arguments mentioned by @Blaisorblade [in Zulip](https://coq.zulipchat.com/#narrow/stream/237656-Coq-devs.20.26.20plugin.20devs/topic/Coq-as-compiler/near/216384164) (indepentently of the macOS config, which currently just sets `native-compiler no` unconditionally),
* and the fact that for Coq < 8.13, your change would have a distracting behavior: with the default option "yes" (`-native-compiler "ondemand"` doesn't exist for these old releases), the stdlib would be precompiled!… which would make macOS-opam installs broken.

So unless someone have a better (albeit more complex) idea, I'd suggest to keep this flag as is in the CEP / item-1

Reply:
Agree, that was just a suggestion.